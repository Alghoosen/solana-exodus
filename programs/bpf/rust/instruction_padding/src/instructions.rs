//! Example Rust-based BPF program that creates large instructions

use {
    solana_program::{
        instruction::{AccountMeta, Instruction},
        program_error::ProgramError,
        pubkey::Pubkey,
    },
    std::{convert::TryInto, mem::size_of},
};

pub fn create_padded_instruction(
    program_id: Pubkey,
    instruction: Instruction,
    padding_accounts: Vec<AccountMeta>,
    padding_data: u32,
) -> Result<Instruction, ProgramError> {
    // The format for instruction data goes:
    // * 4 bytes for the number of accounts required
    // * 4 bytes for the size of the data required
    // * the actual instruction data
    // * additional bytes are all padding
    let mut data = Vec::with_capacity(
        size_of::<u32>() + size_of::<u32>() + instruction.data.len() + padding_data as usize,
    );
    let num_accounts: u32 = instruction
        .accounts
        .len()
        .try_into()
        .map_err(|_| ProgramError::InvalidInstructionData)?;
    data.extend(num_accounts.to_le_bytes().into_iter());

    let data_size: u32 = instruction
        .data
        .len()
        .try_into()
        .map_err(|_| ProgramError::InvalidInstructionData)?;
    data.extend(data_size.to_le_bytes().into_iter());

    data.extend(instruction.data.into_iter());

    for i in 0..padding_data {
        data.push((i % u8::MAX as u32) as u8);
    }

    // The format for account data goes:
    // * first account is the program to call into
    // * following accounts are the one required for the CPI
    // * additional accounts may be included as padding or to test loading / locks
    let mut accounts = Vec::with_capacity(instruction.accounts.len() + 1 + padding_accounts.len());
    accounts.extend(instruction.accounts.into_iter());
    accounts.push(AccountMeta::new_readonly(instruction.program_id, false));
    accounts.extend(padding_accounts.into_iter());

    Ok(Instruction {
        program_id,
        accounts,
        data,
    })
}
