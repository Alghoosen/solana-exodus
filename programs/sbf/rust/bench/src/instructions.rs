//! Example Rust-based SBF program that issues a cross-program-invocation

use {
    core::slice,
    std::{mem, ptr},
};

pub enum BenchInstruction {
    Nop,
    ReadAccounts(ReadAccounts),
    WriteAccounts(WriteAccounts),
    Recurse(Recurse),
}

impl BenchInstruction {
    pub fn tag(&self) -> u8 {
        match self {
            BenchInstruction::Nop => 0,
            BenchInstruction::ReadAccounts(_) => 1,
            BenchInstruction::WriteAccounts(_) => 2,
            BenchInstruction::Recurse(_) => 3,
        }
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = vec![self.tag()];
        match self {
            BenchInstruction::Nop => {}
            BenchInstruction::ReadAccounts(read_accounts) => bytes.extend_from_slice(unsafe {
                slice::from_raw_parts(
                    read_accounts as *const _ as *const u8,
                    mem::size_of::<ReadAccounts>(),
                )
            }),
            BenchInstruction::WriteAccounts(write_accounts) => bytes.extend_from_slice(unsafe {
                slice::from_raw_parts(
                    write_accounts as *const _ as *const u8,
                    mem::size_of::<WriteAccounts>(),
                )
            }),
            BenchInstruction::Recurse(recurse) => bytes.extend_from_slice(unsafe {
                slice::from_raw_parts(recurse as *const _ as *const u8, mem::size_of::<Recurse>())
            }),
        }
        bytes
    }
}

impl TryFrom<&[u8]> for BenchInstruction {
    type Error = ();

    fn try_from(input: &[u8]) -> Result<Self, Self::Error> {
        let (tag, rest) = input.split_first().ok_or(())?;
        match tag {
            0 => Ok(BenchInstruction::Nop),
            1 => {
                let read_accounts =
                    unsafe { ptr::read_unaligned(rest.as_ptr() as *const ReadAccounts) };
                Ok(BenchInstruction::ReadAccounts(read_accounts))
            }
            2 => {
                let write_accounts =
                    unsafe { ptr::read_unaligned(rest.as_ptr() as *const WriteAccounts) };
                Ok(BenchInstruction::WriteAccounts(write_accounts))
            }
            3 => {
                let recurse = unsafe { ptr::read_unaligned(rest.as_ptr() as *const Recurse) };
                Ok(BenchInstruction::Recurse(recurse))
            }
            _ => Err(()),
        }
    }
}

#[repr(C)]
pub struct ReadAccounts {
    pub num_accounts: usize,
    pub size: usize,
}

#[repr(C)]
pub struct WriteAccounts {
    pub num_accounts: usize,
    pub size: usize,
}

#[repr(C)]
pub struct Recurse {
    pub n: usize,
}
