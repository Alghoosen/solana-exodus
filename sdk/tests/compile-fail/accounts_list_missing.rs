use serde_derive::{Deserialize, Serialize};
use solana_sdk_program_macros::instructions;

mod test_program;

#[instructions(test_program::id())]
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub enum TestInstruction {
    #[accounts]
    Transfer { lamports: u64 },
}

fn main() {}
