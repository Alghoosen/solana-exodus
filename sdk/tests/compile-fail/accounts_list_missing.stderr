error: missing accounts list
 --> $DIR/accounts_list_missing.rs:9:7
  |
9 |     #[accounts]
  |       ^^^^^^^^
